### ISSUE TITLE
inference_model is slower than the single_gpu_test.

### ISSUE THREAD

**Issue by BELZHANG** [open-mmlab/mmclassification#802]
I'm trying to use infernce_model function in my own app, it takes about 300ms to output a Res50 result. and with same configuration in the test.py, the speed of each inference becomes normal, both model should run on GPU, and cant figure out which part goes wrong. 

Thanks 
### Post related information
1. The output of `pip list | grep "mmcv\|mmcls\|^torch"`
mmcls              0.22.0   
mmcv-full          1.4.8
torch              1.10.2
torchvision        0.11.3
2. Your config file if you modified it or created a new one.
```python
_base_ = [
    '../_base_/models/resnet50.py', '../_base_/datasets/imagenet_bs32.py',
    '../_base_/schedules/imagenet_bs256.py', '../_base_/default_runtime.py'
]


model = dict(
    head=dict(
        type='LinearClsHead',
        num_classes=12,
        in_channels=2048,
        topk=(1, ),
    ))
data = dict(
    samples_per_gpu=1,
    workers_per_gpu=1,
    train=dict(
        data_prefix='-----------------------------------mydata',
        ann_file='-----------------------------------mydata',
        classes='-----------------------------------mydata',
    val=dict(
        data_prefix='-----------------------------------mydata'
        ann_file='-----------------------------------mydata',
        classes='-----------------------------------mydata'),
    test=dict(
        # replace `data/val` with `data/test` for standard test
        data_prefix='-----------------------------------mydata',
        ann_file='-----------------------------------mydata',
        classes='-----------------------------------mydata'))
evaluation = dict(interval=5, metric='accuracy', metric_options={'topk': (1, )})
3. Your train log file if you meet the problem during training.
0.339 -- speed in s, 
4. Other code you modified in the `mmcls` folder.
      class mm_Predictor:
      
          def __init__(self, cfg, checkpoint):
              self.device = 'cuda:0'
              self.cfg = mmcv.Config.fromfile(cfg)
              self.model = init_model(self.cfg, checkpoint, device=self.device)      
      
          def predict(self, img):
      
              outputs = inference_model(self.model, img)
      
          --------------------------------------------------------------------------------
    
            t0 = time.time()
            result = model(return_loss=False, **data)
            t1 = time.time()
            print('\n')
            print(t1-t0)
      

      this is what my codes look like, and where i set the timer.

**Comment by Ezra-Yu** [comment]
Can you tell us the gap between those methods?

here is the code of [inference_model](https://github.com/open-mmlab/mmclassification/blob/6beac5060f0c1d57141642dba77c19c9ed6b3873/mmcls/apis/inference.py#L54), there is indeed 
 some lines can be improved, but I don't know how big the gap is?

**Comment by BELZHANG** [comment]
I looked the code within the single_gpu_test, looks like it uses similar functions, and it only takes about 0.033s, and I put the timer at the same place.

    for i, data in enumerate(data_loader):
        with torch.no_grad():
            t0 = time.time()
            result = model(return_loss=False, **data)
            t1 = time.time()
            print('\n')
            print(t1-t0)

**Comment by Ezra-Yu** [comment]
So, the gap is 0.033s.
1. you method to caculate time lacks post_processing time, refer to [these lines](https://github.com/open-mmlab/mmclassification/blob/6beac5060f0c1d57141642dba77c19c9ed6b3873/mmcls/apis/inference.py#L84-L88)
2. [inference_model](https://github.com/open-mmlab/mmclassification/blob/6beac5060f0c1d57141642dba77c19c9ed6b3873/mmcls/apis/inference.py#L54) may include reading the image time.
3. [inference_model](https://github.com/open-mmlab/mmclassification/blob/6beac5060f0c1d57141642dba77c19c9ed6b3873/mmcls/apis/inference.py#L54) instantiate pipeline every inference. refer to [lines](https://github.com/open-mmlab/mmclassification/blob/6beac5060f0c1d57141642dba77c19c9ed6b3873/mmcls/apis/inference.py#L65-L78), There should be little impact on time here.

**Comment by BELZHANG** [comment]
Thanks, i believe i didn't include the post-processing time, 
here is the code in single_gpu_test:
def single_gpu_test(model,
                    data_loader,
                    show=False,
                    out_dir=None,
                    **show_kwargs):
    model.eval()
    results = []
    dataset = data_loader.dataset
    prog_bar = mmcv.ProgressBar(len(dataset))
    for i, data in enumerate(data_loader):
        with torch.no_grad():
            t0 = time.time()
            result = model(return_loss=False, **data)
            t1 = time.time()
            print('\n')
            print(t1-t0)


        batch_size = len(result)
        results.extend(result)

        if show or out_dir:
            scores = np.vstack(result)
            pred_score = np.max(scores, axis=1)
            pred_label = np.argmax(scores, axis=1)
            pred_class = [model.CLASSES[lb] for lb in pred_label]

            img_metas = data['img_metas'].data[0]
            imgs = tensor2imgs(data['img'], **img_metas[0]['img_norm_cfg'])
            assert len(imgs) == len(img_metas)

            for i, (img, img_meta) in enumerate(zip(imgs, img_metas)):
                h, w, _ = img_meta['img_shape']
                img_show = img[:h, :w, :]

                ori_h, ori_w = img_meta['ori_shape'][:-1]
                img_show = mmcv.imresize(img_show, (ori_w, ori_h))

                if out_dir:
                    out_file = osp.join(out_dir, img_meta['ori_filename'])
                else:
                    out_file = None

                result_show = {
                    'pred_score': pred_score[i],
                    'pred_label': pred_label[i],
                    'pred_class': pred_class[i]
                }
                model.module.show_result(
                    img_show,
                    result_show,
                    show=show,
                    out_file=out_file,
                    **show_kwargs)

        batch_size = data['img'].size(0)
        for _ in range(batch_size):
            prog_bar.update()
    return results

-----------------------------------------------------
and here is the code from inference_model:

def inference_model(model, img):
    """Inference image(s) with the classifier.

    Args:
        model (nn.Module): The loaded classifier.
        img (str/ndarray): The image filename or loaded image.

    Returns:
        result (dict): The classification results that contains
            `class_name`, `pred_label` and `pred_score`.
    """
    cfg = model.cfg
    device = next(model.parameters()).device  # model device
    # build the data pipeline
    if isinstance(img, str):
        if cfg.data.test.pipeline[0]['type'] != 'LoadImageFromFile':
            cfg.data.test.pipeline.insert(0, dict(type='LoadImageFromFile'))
        data = dict(img_info=dict(filename=img), img_prefix=None)
    else:
        if cfg.data.test.pipeline[0]['type'] == 'LoadImageFromFile':
            cfg.data.test.pipeline.pop(0)
        data = dict(img=img)
    test_pipeline = Compose(cfg.data.test.pipeline)
    data = test_pipeline(data)
    data = collate([data], samples_per_gpu=1)
    if next(model.parameters()).is_cuda:
        # scatter to specified GPU
        data = scatter(data, [device])[0]

    # forward the model
    model.eval()
    with torch.no_grad():
        t0 = time.time()
        scores = model(return_loss=False, **data)
        t1 = time.time()
        print(t1-t0)
    #     pred_score = np.max(scores, axis=1)[0]
    #     pred_label = np.argmax(scores, axis=1)[0]
    #     result = {'pred_label': pred_label, 'pred_score': float(pred_score)}
    # result['pred_class'] = model.CLASSES[result['pred_label']]
    return scores

-------------------------------------------------------------------

and it still give me the time difference, my initial  guess is that I didn't put the model on GPU, nvidia-smi shows the program does use GPU memory while running, and I haven't figure out which step goes wrong.

**Comment by Ezra-Yu** [comment]
please take the average of 100 for comparison.
